dali/operators/reader/loader/discover_files_s3.cc

```diff
@@ -0,0 +1,90 @@
+// Copyright (c) 2017-2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/operators/reader/loader/discover_files_s3.h"
+#include <fnmatch.h>
+#include <filesystem>
+#include <string>
+#include <unordered_map>
+#include <vector>
+#include "dali/operators/reader/loader/discover_files.h"
+#include "dali/util/s3_client_manager.h"
+#include "dali/util/s3_filesystem.h"
+
+namespace dali {
+
+std::vector<FileLabelEntry> s3_discover_files(const std::string &file_root,
+                                              const FileDiscoveryOptions &opts) {
+  assert(starts_with(file_root, "s3://"));
+  auto s3_object_location = s3_filesystem::parse_uri(file_root);
+  std::filesystem::path parent_object_key(s3_object_location.object);
```

Comment:
Well, this is a bit of an UB - using local filesystem path handling to process remote paths. For practical reasons, I don't think we should write utilities to compute relative URI paths, but we can at least require the host system to be UNIX-based.
```
#ifndef __unix__
#error This code works only in UNIX-based systems
#endif
```

Reply:
Done